cmd/terramate/cli/project.go

```diff
@@ -119,35 +100,76 @@ func (p *project) remoteDefaultCommit() string {
 	return p.git.remoteDefaultBranchCommit
 }
 
-func (p *project) isDefaultBranch() bool {
-	git := p.gitcfg()
-	branch, err := p.git.wrapper.CurrentBranch()
-	if err != nil {
-		// WHY?
-		// The current branch name (the symbolic-ref of the HEAD) is not always
-		// available, in this case we naively check if HEAD == local origin/main.
-		// This case usually happens in the git setup of CIs.
-		return p.localDefaultBranchCommit() == p.headCommit()
+// defaultBaseRev returns the revision used for change comparison based on the current Git state.
+func (p *project) defaultBaseRev() string {
+	// Details:
+	// Given origin/main is the default remote/branch, at commit C.
+	// We assume C is the state that ran the last deployment. HEAD is at commit H.
+	//
+	// There's three scenarios, selected if one of the respective cases match, evaluated in order of definition.
+	//
+	//   - Pending changes should be compared to origin/main to find out what has changed since the last deployment.
+	//
+	//     Case 1: H != C and H is not an ancestor of C -- an undeployed, unmerged commit
+	//     Case 2: H == C and symbolic-ref(HEAD) != main -- a new, yet empty branch (=> no changes yet)
+	//
+	//   - Deployed changes should be compared to the previous deployment to find out what changed.
+	//     If we assume that every commit on the main branch is a deployment, that means compare to HEAD^.
+	//
+	//     Case 3: H == C -- latest main commit
+	//     Case 4: H is a first-parent ancestor of main -- previous main commit
+	//
+	//   - Historic changes are all other non-deployed and non-pending, i.e. commits from an already merged and deployed branch.
+	//     They should be compared to the fork point with origin/main.
+	//
+	//     Case 5: H has a fork point with origin/main -- a merged branch commit
+	gitcfg := p.gitcfg()
+	gw := p.git.wrapper
+
+	remoteDefaultBranchRef := p.remoteDefaultBranchRef()
+	headRev, _ := gw.RevParse("HEAD")
```

Comment:
I appreciate the intent of keeping the logic simple here!
Maybe you can put the error handling in a helper (memoized) function.
The error handling is required because there are cases where `HEAD` is not set.

Reply:
same for the other rev-parses

Reply:
The `rev-parse` outputs in the stdout even when an error happens.
Example:
````
➜  terramate git:(snk-git-change-detection) ✗ git rev-parse HEAD:README.md 2>/dev/null
0de873a28b4a0dfe3ef61d369db8085a11b48837
➜  terramate git:(snk-git-change-detection) ✗ echo $?
0
➜  terramate git:(snk-git-change-detection) ✗ git rev-parse HEAD:INVALID.md 2>/dev/null
HEAD:INVALID.md
➜  terramate git:(snk-git-change-detection) ✗ echo $?
128
```

I would do this as a defensive programming because we depend on the git version installed in the user end.

Reply:
All of this error-ignoring relies on `Git.exec()` always returning an empty string in the error case:
```
stdout, err := cmd.Output()
if err != nil {
  stderr := []byte{}
  var exitError *exec.ExitError
  if errors.As(err, &exitError) {
	  stderr = exitError.Stderr
  }
  
  return "", NewCmdError(cmd.String(), stdout, stderr)
}
```
So while the executing rev-parse directly does return the error on stdout, `gw.RevParse` should not.